### PR TITLE
Refactor `Rc` and `Arc` to use a prefix allocator

### PR DESCRIPTION
**Pull request by TimDiekmann**
This picks up #80273 and splits the inner part of `(A)Rc` into two. Both structs are now allocated as if they were allocated with a prefix allocator, which may be useful in the future. For now, the allocator is gated behind `allocator_api_internals`.

The `Tracker`-allocator is used in tests to ensure that the safety rules are not violated.

r? @Amanieu
This requires a perf-run.

Update: 2021/04/22: With this, `Rc` and `Arc` now stores the pointer returned from the allocator, i.e. it points to `T` directly because accessing the data in smart pointers is more common than ~copying~ cloning the smart pointer. The layout of the allocation didn't change.

### CODE REVIEW

**Comment by TimDiekmann** [comment]
@rust-lang/infra CI is using `GNU gdb (Ubuntu 8.1.1-0ubuntu1) 8.1.1` which is missing `gdb.Type.alignof` so it's not possible to retrieve the alignment of a type. Is there a specific reason for using an old version of gdb? The latest version at the time of writing is 10.1.

File: library/alloc/src/sync.rs [hunk 1]
```diff
@@ -239,12 +271,9 @@ impl<T: ?Sized + Unsize<U>, U: ?Sized> CoerceUnsized<Arc<U>> for Arc<T> {}
 impl<T: ?Sized + Unsize<U>, U: ?Sized> DispatchFromDyn<Arc<U>> for Arc<T> {}
 
 impl<T: ?Sized> Arc<T> {
-    fn from_inner(ptr: NonNull<ArcInner<T>>) -> Self {
-        Self { ptr, phantom: PhantomData }
-    }
-
-    unsafe fn from_ptr(ptr: *mut ArcInner<T>) -> Self {
-        unsafe { Self::from_inner(NonNull::new_unchecked(ptr)) }
+    #[inline(always)]
```
[the remark this answers is not in the document]
Comment (TimDiekmann) [reply library/alloc/src/sync.rs:274]: I copied `#[inline(always)]` from `fn inner`. I'll change it to `#[inline]` as the function got more complicated.

File: library/alloc/src/sync.rs [hunk 2]
```diff
@@ -326,15 +349,8 @@ impl<T> Arc<T> {
...
-        // Start the weak pointer count as 1 which is the weak pointer that's
-        // held by all the strong pointers (kinda), see std/rc.rs for more info
-        let x: Box<_> = box ArcInner {
-            strong: atomic::AtomicUsize::new(1),
-            weak: atomic::AtomicUsize::new(1),
-            data,
-        };
-        Self::from_inner(Box::leak(x).into())
+    pub fn new(data: T) -> Self {
+        Self::try_new(data).unwrap_or_else(|_| handle_alloc_error(Layout::new::<T>()))
```
[the remark this answers is not in the document]
Comment (TimDiekmann) [reply library/alloc/src/sync.rs:353]: I moved the call to the allocator (and to `handle_alloc_error`) out of `NonNull`.

**Comment by Amanieu** [comment]
I have no idea why this specific unrelated test is failing all of a sudden...

**Comment by Amanieu** [comment]
This is most likely #84025

**Comment by TimDiekmann** [comment]
I have no idea, either. So we wait until the issue is resolved and r+ again?

Do we want to do another perf-run with the latest changes?

**Comment by bjorn3** [comment]
Large regressions up to 20%.

**Comment by crlf0710** [comment]
@TimDiekmann Triage: what's next steps here?

**Comment by TimDiekmann** [comment]
> @TimDiekmann Triage: what's next steps here?

Honestly, I do not know. @Amanieu Could you look at the code again? Maybe you can find something that explains the big regressions? Or this is exclusively the new logic as other tests got better in performance?

**Comment by Amanieu** [comment]
I would recommend using [cargo-llvm-lines](https://github.com/dtolnay/cargo-llvm-lines) to see which generic functions are being instantiated too much and focus on optimizing those. You can download the try build from bors (ae3baaa934d54cb4af3f0bb658e735a04b77215b) using [rustup-toolchain-install-master](https://github.com/kennytm/rustup-toolchain-install-master).

**Comment by TimDiekmann** [comment]
> I would recommend using [cargo-llvm-lines](https://github.com/dtolnay/cargo-llvm-lines) to see which generic functions are being instantiated too much and focus on optimizing those. You can download the try build from bors ([ae3baaa](https://github.com/rust-lang/rust/commit/ae3baaa934d54cb4af3f0bb658e735a04b77215b)) using [rustup-toolchain-install-master](https://github.com/kennytm/rustup-toolchain-install-master).

Thanks, will take a look!

**Comment by TimDiekmann** [comment]
Comparing the generated llvm lines of [ae3baaa](https://github.com/rust-lang/rust/commit/ae3baaa934d54cb4af3f0bb658e735a04b77215b) with it's parent [76a04dd](https://github.com/rust-lang/rust/commit/76a04dd3f8b2500d9e524dca122814fd5a3e25e4) for the [regression-31157 benchmark](https://github.com/rust-lang/rustc-perf/tree/master/collector/benchmarks/regression-31157) (which has the highest regression), those are the results:

```
$ cargo +76a04dd3f8b2500d9e524dca122814fd5a3e25e4 llvm-lines -q 2> /dev/null | grep 'alloc::' | head -20
   4049 (6.4%)    67 (2.5%)  alloc::alloc::box_free
   2536 (4.0%)    39 (1.5%)  alloc::rc::Rc<T>::new
   1700 (2.7%)    39 (1.5%)  <alloc::rc::Rc<T> as core::ops::drop::Drop>::drop
   1131 (1.8%)    39 (1.5%)  core::alloc::layout::Layout::for_value
    978 (1.6%)     6 (0.2%)  alloc::raw_vec::RawVec<T,A>::grow_amortized
    784 (1.2%)    28 (1.1%)  alloc::rc::RcInnerPtr::inc_strong
    590 (0.9%)    40 (1.5%)  alloc::boxed::Box<T,A>::leak
    588 (0.9%)     4 (0.2%)  alloc::raw_vec::RawVec<T,A>::allocate_in
    576 (0.9%)     8 (0.3%)  core::alloc::layout::Layout::array
    567 (0.9%)     4 (0.2%)  <T as alloc::slice::hack::ConvertVec>::to_vec
    528 (0.8%)     8 (0.3%)  alloc::raw_vec::RawVec<T,A>::current_memory
    468 (0.7%)    20 (0.8%)  <alloc::boxed::Box<F,A> as core::ops::function::Fn<Args>>::call
    429 (0.7%)    39 (1.5%)  alloc::rc::RcInnerPtr::dec_strong
    429 (0.7%)    39 (1.5%)  alloc::rc::RcInnerPtr::dec_weak
    336 (0.5%)     6 (0.2%)  alloc::vec::Vec<T,A>::push
    312 (0.5%)    39 (1.5%)  alloc::rc::Rc<T>::from_inner
    312 (0.5%)    39 (1.5%)  alloc::rc::RcInnerPtr::strong
    312 (0.5%)    39 (1.5%)  alloc::rc::RcInnerPtr::weak
    312 (0.5%)     8 (0.3%)  <alloc::raw_vec::RawVec<T,A> as core::ops::drop::Drop>::drop
    308 (0.5%)    28 (1.1%)  <alloc::rc::Rc<T> as core::clone::Clone>::clone
```

```
$ cargo +ae3baaa934d54cb4af3f0bb658e735a04b77215b llvm-lines --release -q 2> /dev/null | grep 'alloc::' | head -20
   3299 (4.1%)    78 (2.0%)  <alloc::rc::Rc<T> as core::ops::drop::Drop>::drop
   2597 (3.2%)    39 (1.0%)  alloc::rc::Rc<T>::new
   2256 (2.8%)    34 (0.9%)  alloc::alloc::box_free
   2160 (2.7%)    80 (2.0%)  core::alloc::helper::PrefixAllocator<Alloc,Prefix>::prefix
   1920 (2.4%)    80 (2.0%)  core::alloc::layout::Layout::for_value_raw
   1755 (2.2%)    39 (1.0%)  alloc::rc::Rc<T>::allocate
   1482 (1.8%)    26 (0.7%)  alloc::raw_vec::RawVec<T,A>::current_memory
   1014 (1.3%)    39 (1.0%)  alloc::rc::Rc<T>::new_uninit
   1008 (1.3%)     7 (0.2%)  alloc::raw_vec::RawVec<T,A>::grow_amortized
    936 (1.2%)    78 (2.0%)  alloc::rc::Rc<T>::metadata
    832 (1.0%)    52 (1.3%)  core::alloc::layout::size_align
    828 (1.0%)     9 (0.2%)  core::alloc::layout::Layout::array
    806 (1.0%)    26 (0.7%)  <alloc::raw_vec::RawVec<T,A> as core::ops::drop::Drop>::drop
    780 (1.0%)    39 (1.0%)  alloc::rc::Rc<core::mem::maybe_uninit::MaybeUninit<T>>::assume_init
    710 (0.9%)     5 (0.1%)  alloc::raw_vec::RawVec<T,A>::allocate_in
    624 (0.8%)    78 (2.0%)  alloc::rc::Rc<T>::from_raw_in
    616 (0.8%)     5 (0.1%)  <T as alloc::slice::hack::ConvertVec>::to_vec
    572 (0.7%)    52 (1.3%)  core::alloc::layout::Layout::new
    468 (0.6%)    20 (0.5%)  <alloc::boxed::Box<F,A> as core::ops::function::Fn<Args>>::call
    342 (0.4%)     6 (0.2%)  alloc::vec::Vec<T,A>::push
```

Most notable is, that `drop` was called twice as often, while the copies of `box_free` were halved. I'm currently investigating, why this happened, probably because of a temporary `Rc` anywhere. Also, calling `prefix` occurred quite often, probably because of calling `drop`. I think this can be reduced by caching the `Metadata`. For comparison: `NonNull::cast` had ~300 lines in the parent commit. I can't explain however why `call_once` was called three times more. I'll build rustc several times to test different results and hope I can resolve this. If anyone has a hint for me, I would appreciate it very much.

**Comment by TimDiekmann** [comment]
I was able to solve the regression: Calling `new_uninit` from `new` generates all the code twice, once for `MaybeUninit<T>` and once for `T`, even if `Rc<MaybeUninit<T>>` isn't used besides `assume_init`.

```
$ cargo clean && cargo +stage1 llvm-lines -q --release 2> /dev/null | grep 'alloc::' | head -20
   2256 (3.2%)    34 (1.1%)  alloc::alloc::box_free
   2129 (3.0%)    39 (1.2%)  alloc::rc::Rc<T>::new
   1677 (2.4%)    39 (1.2%)  alloc::rc::Rc<T>::allocate
   1505 (2.1%)    39 (1.2%)  <alloc::rc::Rc<T> as core::ops::drop::Drop>::drop
   1482 (2.1%)    26 (0.8%)  alloc::raw_vec::RawVec<T,A>::current_memory
    984 (1.4%)    41 (1.3%)  core::alloc::layout::Layout::for_value_raw
    952 (1.3%)     7 (0.2%)  alloc::raw_vec::RawVec<T,A>::grow_amortized
    832 (1.2%)    52 (1.6%)  core::alloc::layout::size_align
    810 (1.1%)     9 (0.3%)  core::alloc::layout::Layout::array
    806 (1.1%)    26 (0.8%)  <alloc::raw_vec::RawVec<T,A> as core::ops::drop::Drop>::drop
    710 (1.0%)     5 (0.2%)  alloc::raw_vec::RawVec<T,A>::allocate_in
    616 (0.9%)     5 (0.2%)  <T as alloc::slice::hack::ConvertVec>::to_vec
    572 (0.8%)    52 (1.6%)  core::alloc::layout::Layout::new
    468 (0.7%)    39 (1.2%)  alloc::rc::Rc<T>::metadata
    468 (0.7%)    20 (0.6%)  <alloc::boxed::Box<F,A> as core::ops::function::Fn<Args>>::call
    351 (0.5%)    39 (1.2%)  alloc::rc::Rc<T>::metadata_ptr
    342 (0.5%)     6 (0.2%)  alloc::vec::Vec<T,A>::push
    312 (0.4%)    39 (1.2%)  alloc::rc::Rc<T>::from_raw_in
    306 (0.4%)     2 (0.1%)  core::alloc::helper::PrefixAllocator<Alloc,Prefix>::alloc_impl
    280 (0.4%)    28 (0.9%)  <alloc::rc::Rc<T> as core::clone::Clone>::clone
```

Parent commit: 70590 Lines, 3253 Copies
Current commit: 80395 Lines, 3966 Copies
"Fixed" commit: 70861 Lines, 3221 Copies

I'll apply those changes to other methods besides `Rc::new` as well and push the commit so another benchmarkl can be run.

**Comment by TimDiekmann** [comment]
The main regression comes from the optimizing path in LLVM (`LLVM_module_optimize_module_passes` and `LLVM_lto_optimize`). I was able to help the compiler out and simplify the critical sections. I'll push another commit later. 

```
$ cargo +stage1 llvm-lines -q --release 2> /dev/null | grep 'alloc::' | head -20
   2256 (3.3%)    34 (1.1%)  alloc::alloc::box_free
   1856 (2.7%)    39 (1.3%)  alloc::rc::Rc<T>::new
   1505 (2.2%)    39 (1.3%)  <alloc::rc::Rc<T> as core::ops::drop::Drop>::drop
   1482 (2.2%)    26 (0.8%)  alloc::raw_vec::RawVec<T,A>::current_memory
    984 (1.4%)    41 (1.3%)  core::alloc::layout::Layout::for_value_raw
    952 (1.4%)     7 (0.2%)  alloc::raw_vec::RawVec<T,A>::grow_amortized
    832 (1.2%)    52 (1.7%)  core::alloc::layout::size_align
    810 (1.2%)     9 (0.3%)  core::alloc::layout::Layout::array
    806 (1.2%)    26 (0.8%)  <alloc::raw_vec::RawVec<T,A> as core::ops::drop::Drop>::drop
    710 (1.0%)     5 (0.2%)  alloc::raw_vec::RawVec<T,A>::allocate_in
    616 (0.9%)     5 (0.2%)  <T as alloc::slice::hack::ConvertVec>::to_vec
    572 (0.8%)    52 (1.7%)  core::alloc::layout::Layout::new
    468 (0.7%)    39 (1.3%)  alloc::rc::Rc<T>::metadata
    468 (0.7%)    20 (0.6%)  <alloc::boxed::Box<F,A> as core::ops::function::Fn<Args>>::call
    351 (0.5%)    39 (1.3%)  alloc::rc::Rc<T>::metadata_ptr
    342 (0.5%)     6 (0.2%)  alloc::vec::Vec<T,A>::push
    280 (0.4%)    28 (0.9%)  <alloc::rc::Rc<T> as core::clone::Clone>::clone
    275 (0.4%)    26 (0.8%)  <alloc::vec::Vec<T,A> as core::ops::drop::Drop>::drop
    234 (0.3%)    39 (1.3%)  alloc::rc::Rc<T>::from_non_null
    234 (0.3%)    26 (0.8%)  alloc::vec::Vec<T,A>::as_mut_ptr
```

Lines: 68053, Copies: 3117

**Comment by CAD97** [comment]
It's even now, and even positive on the #31157 stress regression test! :tada:

**Comment by CAD97** [comment]
(I'm not smart and read the summary wrong)

**Comment by Amanieu** [comment]
This is still a pretty significant compilation time regression. I don't think we can land this as it is.

**Comment by JohnCSimon** [comment]
Ping from triage: 
@TimDiekmann moving this back to author

@Rustbot label: -S-waiting-on-review +S-waiting-on-author

**Comment by crlf0710** [comment]
@TimDiekmann Ping from triage, any updates on this?

**Comment by TimDiekmann** [comment]
Sadly I wasn't able to make further optimizations, so I close that for now.

**Pull request closed without merging**